pandas/core/indexes/datetimelike.py

```diff
@@ -88,9 +80,51 @@ class DatetimeIndexOpsMixin(NDArrayBackedExtensionIndex):
     _is_numeric_dtype = False
     _can_hold_strings = False
     _data: DatetimeArray | TimedeltaArray | PeriodArray
-    freq: BaseOffset | None
-    freqstr: str | None
-    _resolution_obj: Resolution
+
+    # ------------------------------------------------------------------------
+
+    @doc(DatetimeLikeArrayMixin.mean)
+    def mean(self, *, skipna: bool = True, axis: int | None = 0):
+        return self._data.mean(skipna=skipna, axis=axis)
+
+    # error: Decorated property not supported
+    @property  # type: ignore[misc]
+    @doc(DatetimeLikeArrayMixin.asi8)
+    def asi8(self) -> npt.NDArray[np.int64]:
+        return self._data.asi8
+
+    # error: Decorated property not supported
+    @property  # type: ignore[misc]
+    @doc(DatetimeLikeArrayMixin.freq)
+    def freq(self) -> BaseOffset | None:
+        return self._data.freq
+
+    @freq.setter
+    def freq(self, value) -> None:
+        # error: Property "freq" defined in "PeriodArray" is read-only
+        self._data.freq = value  # type: ignore[misc]
+
+    # error: Decorated property not supported
+    @property  # type: ignore[misc]
+    @doc(DatetimeLikeArrayMixin.freqstr)
+    def freqstr(self) -> str | None:
+        return self._data.freqstr
+
+    # error: Decorated property not supported
+    @cache_readonly  # type: ignore[misc]
+    @doc(DatetimeLikeArrayMixin.inferred_freq)
+    def inferred_freq(self) -> str | None:
```

Comment:
i think this is only for DTI/TDI, not PI?

Reply:
On 1.4.3 (and on this PR) it errors for PeriodIndex:

```py
>>> pd.PeriodIndex(year=[2000, 2002], quarter=[1, 3]).inferred_freq
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/_libs/properties.pyx", line 37, in pandas._libs.properties.CachedProperty.__get__
  File "pandas/core/indexes/extension.py", line 59, in cached
    return getattr(self._data, name)
  File "pandas/core/arrays/datetimelike.py", line 917, in inferred_freq
    return frequencies.infer_freq(self)
  File "pandas/tseries/frequencies.py", line 183, in infer_freq
    raise TypeError(
TypeError: PeriodIndex given. Check the `freq` attribute instead of using infer_freq.
```

Could add a `NoReturn`-annotation to `freqstr` in `PeriodIndex`.

Reply:
yah, so better to just move this to DatetimeTimedeltaMixin?

Reply:
Thank you! I moved it.